Merge more symbolic meta kernels and symint changes from branch

bdhirsh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86334

symintify split_with_sizes, dropout, fused_fake_obs_quant. meta for padding_2d ops

add meta_bernoulli_

meta kernel for at::gather

get pytorch_struct to pass: meta for scatter_add, fix backward

symintify split ops

bdhirsh committed:
symintify split_with_sizes, dropout, fused_fake_obs_quant. meta for padding_2d ops

add meta_bernoulli_

meta kernel for at::gather

get pytorch_struct to pass: meta for scatter_add, fix backward

symintify split ops

[ghstack-poisoned]

ezyang reviewed (approved):
It would be nice if we had test_meta.py tests on register_dispatcher=False regs

bdhirsh committed:
Update on "Merge more symbolic meta kernels and symint changes from branch"


symintify split_with_sizes, dropout, fused_fake_obs_quant. meta for padding_2d ops

add meta_bernoulli_

meta kernel for at::gather

get pytorch_struct to pass: meta for scatter_add, fix backward

symintify split ops

[ghstack-poisoned]

bdhirsh commented:
> It would be nice if we had test_meta.py tests on register_dispatcher=False regs

Coming up here :) https://github.com/pytorch/pytorch/pull/86228

A bunch of errors actually came out of it, so I'm starting to sift through them

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1840,17 +1840,18 @@ Tensor split_with_sizes_backward(
   int64_t num_splits = grads.size();
-  std::vector<int64_t> split_sizes(num_splits, split_size);
+  std::vector<c10::SymInt> split_sizes(num_splits, split_size);
+  auto sym_split_size = c10::SymInt(split_size);
```
It is already a SymInt

bdhirsh committed:
Update on "Merge more symbolic meta kernels and symint changes from branch"


symintify split_with_sizes, dropout, fused_fake_obs_quant. meta for padding_2d ops

add meta_bernoulli_

meta kernel for at::gather

get pytorch_struct to pass: meta for scatter_add, fix backward

symintify split ops

[ghstack-poisoned]

bdhirsh committed:
Update on "Merge more symbolic meta kernels and symint changes from branch"


symintify split_with_sizes, dropout, fused_fake_obs_quant. meta for padding_2d ops

add meta_bernoulli_

meta kernel for at::gather

get pytorch_struct to pass: meta for scatter_add, fix backward

symintify split ops

[ghstack-poisoned]

bdhirsh committed:
Update on "Merge more symbolic meta kernels and symint changes from branch"


symintify split_with_sizes, dropout, fused_fake_obs_quant. meta for padding_2d ops

add meta_bernoulli_

meta kernel for at::gather

get pytorch_struct to pass: meta for scatter_add, fix backward

symintify split ops

[ghstack-poisoned]

bdhirsh commented on torch/_meta_registrations.py:
```diff
@@ -1472,18 +1472,25 @@ def meta_scatter_add(self, dim, index, src):
+    output_width = out_size[1]
+    nbatch = input_shape[0]
+    channels = input_shape[1]
+    return input.new_empty((nbatch, channels, output_height, output_width)).to(memory_format=mem_format)
```
cc @Krovatkin 

tbh I just started at the C++ meta kernel defined in `UpSampleNearest2d.cpp` to get this change, and it passed that failing `test/test_meta.py` test

bdhirsh committed:
Update on "Merge more symbolic meta kernels and symint changes from branch"


symintify split_with_sizes, dropout, fused_fake_obs_quant. meta for padding_2d ops

add meta_bernoulli_

meta kernel for at::gather

get pytorch_struct to pass: meta for scatter_add, fix backward

symintify split ops

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge -g

seemethere commented:
@pytorchbot revert -c ghfirst "Trying to revert https://github.com/pytorch/pytorch/pull/86207, this PR causes merge conflicts with the initial revert so will have to revert this as well"

seemethere commented:
@pytorchbot revert -c ghfirst -m "Trying to revert https://github.com/pytorch/pytorch/pull/86207, this PR causes merge conflicts with the initial revert so will have to revert this as well"

albanD commented:
reland is at https://github.com/pytorch/pytorch/pull/86488
